<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Импорт-данных" data-toc-modified-id="Импорт-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Импорт данных</a></span></li><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Предобработка" data-toc-modified-id="Предобработка-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Предобработка</a></span></li><li><span><a href="#Кодирование-категориальных-признаков" data-toc-modified-id="Кодирование-категориальных-признаков-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Кодирование категориальных признаков</a></span></li></ul></li><li><span><a href="#Исследование-баланса-классов" data-toc-modified-id="Исследование-баланса-классов-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследование баланса классов</a></span></li><li><span><a href="#Обучение-модели" data-toc-modified-id="Обучение-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Обучение модели</a></span><ul class="toc-item"><li><span><a href="#Кодирование-категориальных-переменных" data-toc-modified-id="Кодирование-категориальных-переменных-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Кодирование категориальных переменных</a></span></li><li><span><a href="#Стандартизация-данных" data-toc-modified-id="Стандартизация-данных-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Стандартизация данных</a></span></li><li><span><a href="#Обучение-без-учета-дисбаланса-классов" data-toc-modified-id="Обучение-без-учета-дисбаланса-классов-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Обучение без учета дисбаланса классов</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-4.3.1"><span class="toc-item-num">4.3.1&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-4.3.2"><span class="toc-item-num">4.3.2&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-4.3.3"><span class="toc-item-num">4.3.3&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li></ul></li><li><span><a href="#Учтем-дисбаланс-классов" data-toc-modified-id="Учтем-дисбаланс-классов-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Учтем дисбаланс классов</a></span><ul class="toc-item"><li><span><a href="#Увеличение-выборки" data-toc-modified-id="Увеличение-выборки-4.4.1"><span class="toc-item-num">4.4.1&nbsp;&nbsp;</span>Увеличение выборки</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-4.4.1.1"><span class="toc-item-num">4.4.1.1&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-4.4.1.2"><span class="toc-item-num">4.4.1.2&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-4.4.1.3"><span class="toc-item-num">4.4.1.3&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li></ul></li><li><span><a href="#Уменьшение-выборки" data-toc-modified-id="Уменьшение-выборки-4.4.2"><span class="toc-item-num">4.4.2&nbsp;&nbsp;</span>Уменьшение выборки</a></span></li><li><span><a href="#Изменение-порога" data-toc-modified-id="Изменение-порога-4.4.3"><span class="toc-item-num">4.4.3&nbsp;&nbsp;</span>Изменение порога</a></span></li></ul></li></ul></li><li><span><a href="#Тестирование-лучшей-модели" data-toc-modified-id="Тестирование-лучшей-модели-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Тестирование лучшей модели</a></span></li><li><span><a href="#Общий-вывод-по-проекту" data-toc-modified-id="Общий-вывод-по-проекту-6"><span class="toc-item-num">6&nbsp;&nbsp;</span><b>Общий вывод по проекту</b></a></span></li><li><span><a href="#Общий-вывод-по-проекту-В2" data-toc-modified-id="Общий-вывод-по-проекту-В2-7"><span class="toc-item-num">7&nbsp;&nbsp;</span><b>Общий вывод по проекту В2</b></a></span></li></ul></div>

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.
Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Требуется построить модель с предельно большим значением F1-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59.

## Импорт данных

In [1]:
import pandas as pd

from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score, roc_auc_score
from sklearn.utils import shuffle


In [2]:
try:
    data = pd.read_csv('C:\datasets\Churn.csv')
except:
    data = pd.read_csv('/datasets/Churn.csv')


In [3]:
data.head(10)

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   
5          6    15574012       Chu          645     Spain    Male   44   
6          7    15592531  Bartlett          822    France    Male   50   
7          8    15656148    Obinna          376   Germany  Female   29   
8          9    15792365        He          501    France    Male   44   
9         10    15592389        H?          684    France    Male   27   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   
5     8.0  113755.78              2          1               0   
6     7.0       0.00              2          1               1   
7     4.0  115046.74              4          1               0   
8     4.0  142051.07              2          0               1   
9     2.0  134603.88              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0  
5        149756.71       1  
6         10062.80       0  
7        119346.88       1  
8         74940.50       0  
9         71725.73       0

In [4]:
data.tail(10)

RowNumber  CustomerId      Surname  CreditScore Geography  Gender  Age  \
9990       9991    15798964   Nkemakonam          714   Germany    Male   33   
9991       9992    15769959  Ajuluchukwu          597    France  Female   53   
9992       9993    15657105  Chukwualuka          726     Spain    Male   36   
9993       9994    15569266       Rahman          644    France    Male   28   
9994       9995    15719294         Wood          800    France  Female   29   
9995       9996    15606229     Obijiaku          771    France    Male   39   
9996       9997    15569892    Johnstone          516    France    Male   35   
9997       9998    15584532          Liu          709    France  Female   36   
9998       9999    15682355    Sabbatini          772   Germany    Male   42   
9999      10000    15628319       Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
9990     3.0   35016.60              1          1               0   
9991     4.0   88381.21              1          1               0   
9992     2.0       0.00              1          1               0   
9993     7.0  155060.41              1          1               0   
9994     2.0       0.00              2          0               0   
9995     5.0       0.00              2          1               0   
9996    10.0   57369.61              1          1               1   
9997     7.0       0.00              1          0               1   
9998     3.0   75075.31              2          1               0   
9999     NaN  130142.79              1          1               0   

      EstimatedSalary  Exited  
9990         53667.08       0  
9991         69384.71       1  
9992        195192.40       0  
9993         29179.52       0  
9994        167773.55       0  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


В датасете есть пропуски, не все типы данных оптимальны для значений в столбцах.

## Подготовка данных

### Предобработка

Столбец `Row Number` аналогичен столбцу индексов, поэтому от него можно избавиться

In [6]:
data = data.drop(['RowNumber'], axis=1)
data.head()

CustomerId   Surname  CreditScore Geography  Gender  Age  Tenure  \
0    15634602  Hargrave          619    France  Female   42     2.0   
1    15647311      Hill          608     Spain  Female   41     1.0   
2    15619304      Onio          502    France  Female   42     8.0   
3    15701354      Boni          699    France  Female   39     1.0   
4    15737888  Mitchell          850     Spain  Female   43     2.0   

     Balance  NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  \
0       0.00              1          1               1        101348.88   
1   83807.86              1          0               1        112542.58   
2  159660.80              3          1               0        113931.57   
3       0.00              2          0               0         93826.63   
4  125510.82              1          1               1         79084.10   

   Exited  
0       1  
1       0  
2       1  
3       0  
4       0

Столбцы с Фамилией `Surname` и с id `CustomerId` никак нам не помогут в обучении модели, поэтому от них тоже избавимся

In [7]:
data = data.drop(['Surname'], axis=1)
data = data.drop(['CustomerId'], axis=1)
data.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42     2.0       0.00              1   
1          608     Spain  Female   41     1.0   83807.86              1   
2          502    France  Female   42     8.0  159660.80              3   
3          699    France  Female   39     1.0       0.00              2   
4          850     Spain  Female   43     2.0  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1        101348.88       1  
1          0               1        112542.58       0  
2          1               0        113931.57       1  
3          0               0         93826.63       0  
4          1               1         79084.10       0

Приведем названия столбцов к нижнему 'змеиному' регистру.

In [8]:
data.columns = [
    'credit_score',
    'geography',
    'gender',
    'age',
    'tenure',
    'balance',
    'num_of_products',
    'has_cr_card',
    'is_active_member',
    'estimated_salary',
    'exited'
]
data.head()

credit_score geography  gender  age  tenure    balance  num_of_products  \
0           619    France  Female   42     2.0       0.00                1   
1           608     Spain  Female   41     1.0   83807.86                1   
2           502    France  Female   42     8.0  159660.80                3   
3           699    France  Female   39     1.0       0.00                2   
4           850     Spain  Female   43     2.0  125510.82                1   

   has_cr_card  is_active_member  estimated_salary  exited  
0            1                 1         101348.88       1  
1            0                 1         112542.58       0  
2            1                 0         113931.57       1  
3            0                 0          93826.63       0  
4            1                 1          79084.10       0

In [9]:
data['credit_score'].max()

850

Преобразуем столбец `credit_score` в более экономичный с точки зрения памяти тип данных `int16`, столбец `age` в `int8`

In [10]:
data['credit_score'] = data['credit_score'].astype('int16')

In [11]:
data['age'].max()

92

In [12]:
data['age'] = data['age'].astype('int8')

In [13]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   credit_score      10000 non-null  int16  
 1   geography         10000 non-null  object 
 2   gender            10000 non-null  object 
 3   age               10000 non-null  int8   
 4   tenure            9091 non-null   float64
 5   balance           10000 non-null  float64
 6   num_of_products   10000 non-null  int64  
 7   has_cr_card       10000 non-null  int64  
 8   is_active_member  10000 non-null  int64  
 9   estimated_salary  10000 non-null  float64
 10  exited            10000 non-null  int64  
dtypes: float64(3), int16(1), int64(4), int8(1), object(2)
memory usage: 732.5+ KB


In [14]:
data['tenure'].value_counts()

1.0     952
2.0     950
8.0     933
3.0     928
5.0     927
7.0     925
4.0     885
9.0     882
6.0     881
10.0    446
0.0     382
Name: tenure, dtype: int64

In [15]:
data['tenure'].isna().sum()

909

In [16]:
data[data['tenure'].isna()]

credit_score geography  gender  age  tenure    balance  num_of_products  \
30             591     Spain  Female   39     NaN       0.00                3   
48             550   Germany    Male   38     NaN  103391.38                1   
51             585   Germany    Male   36     NaN  146050.97                2   
53             655   Germany    Male   41     NaN  125561.97                1   
60             742   Germany    Male   35     NaN  136857.00                1   
...            ...       ...     ...  ...     ...        ...              ...   
9944           744   Germany    Male   41     NaN  190409.34                2   
9956           520    France  Female   46     NaN   85216.61                1   
9964           479    France    Male   34     NaN  117593.48                2   
9985           659    France    Male   36     NaN  123841.49                2   
9999           792    France  Female   28     NaN  130142.79                1   

      has_cr_card  is_active_member  estimated_salary  exited  
30              1                 0         140469.38       1  
48              0                 1          90878.13       0  
51              0                 0          86424.57       0  
53              0                 0         164040.94       1  
60              0                 0          84509.57       0  
...           ...               ...               ...     ...  
9944            1                 1         138361.48       0  
9956            1                 0         117369.52       1  
9964            0                 0         113308.29       0  
9985            1                 0          96833.00       0  
9999            1                 0          38190.78       0  

[909 rows x 11 columns]

Скорее всего пропуски появились из-за технической ошибки или незаполненности данных нулями. Заполним значения NaN медианой.

In [17]:
data['tenure'] = data['tenure'].fillna(value=data['tenure'].median())

In [18]:
data.isna().sum()

credit_score        0
geography           0
gender              0
age                 0
tenure              0
balance             0
num_of_products     0
has_cr_card         0
is_active_member    0
estimated_salary    0
exited              0
dtype: int64

Столбец tenure (сколько лет человек является клиентом банка) содержит в себе только целочисленные значения, а значит и его можно преобразовать к более эффективному типу данных.

In [19]:
data['tenure'] = data['tenure'].astype('int8')

In [20]:
data['num_of_products'].max()

4

In [21]:
data['num_of_products'] = data['num_of_products'].astype('int8')

Столбец `has_a_card` содержит в себе только значения 1 или 0, заменим его тип данных на `int8` (хотя можем и на bool, с точки зрения памяти - разницы не будет, с математической точки зрения для python тоже разницы нет). Аналогично и столбец `is_active_member` и `exited`.

In [22]:
data['has_cr_card'] = data['has_cr_card'].astype('int8')

In [23]:
data['is_active_member'] = data['is_active_member'].astype('int8')

In [24]:
data['exited'] = data['exited'].astype('int8')

Проверим на наличие дубликатов

In [25]:
data.duplicated().sum()

0

In [26]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   credit_score      10000 non-null  int16  
 1   geography         10000 non-null  object 
 2   gender            10000 non-null  object 
 3   age               10000 non-null  int8   
 4   tenure            10000 non-null  int8   
 5   balance           10000 non-null  float64
 6   num_of_products   10000 non-null  int8   
 7   has_cr_card       10000 non-null  int8   
 8   is_active_member  10000 non-null  int8   
 9   estimated_salary  10000 non-null  float64
 10  exited            10000 non-null  int8   
dtypes: float64(2), int16(1), int8(6), object(2)
memory usage: 390.8+ KB


Удалось сократить использование памяти более чем в 2 раза.

## Исследование баланса классов

У нас было несколько категориальных переменных - gender, geography, has_cr_card, is_active_member, exited. Посмотрим на их распределения

In [29]:
data['exited'].value_counts()

0    7963
1    2037
Name: exited, dtype: int64

Классы в `exited` имеют сильный дисбаланс (соотношение 4:1).

## Обучение модели

Требуется получить модель с F1-мерой >= 0.59

### Кодирование категориальных переменных

In [30]:
target = data['exited']
features = data.drop('exited', axis=1)
features

credit_score geography  gender  age  tenure    balance  num_of_products  \
0              619    France  Female   42       2       0.00                1   
1              608     Spain  Female   41       1   83807.86                1   
2              502    France  Female   42       8  159660.80                3   
3              699    France  Female   39       1       0.00                2   
4              850     Spain  Female   43       2  125510.82                1   
...            ...       ...     ...  ...     ...        ...              ...   
9995           771    France    Male   39       5       0.00                2   
9996           516    France    Male   35      10   57369.61                1   
9997           709    France  Female   36       7       0.00                1   
9998           772   Germany    Male   42       3   75075.31                2   
9999           792    France  Female   28       5  130142.79                1   

      has_cr_card  is_active_member  estimated_salary  
0               1                 1         101348.88  
1               0                 1         112542.58  
2               1                 0         113931.57  
3               0                 0          93826.63  
4               1                 1          79084.10  
...           ...               ...               ...  
9995            1                 0          96270.64  
9996            1                 1         101699.77  
9997            0                 1          42085.58  
9998            1                 0          92888.52  
9999            1                 0          38190.78  

[10000 rows x 10 columns]

Разделяем выборку на 3 части: обучающую, тестовую и валидационную в соотношениях 3:1:1.

In [31]:
features_train, features_tv, target_train, target_tv = train_test_split(
    features, target, test_size=0.4, random_state=12345)

print('train:',features_train.shape, target_train.shape)

features_test, features_valid, target_test, target_valid = train_test_split(
    features_tv, target_tv, test_size=0.5, random_state=12345)

print('train:',features_test.shape, target_test.shape)
print('train:',features_valid.shape, target_valid.shape)

train: (6000, 10) (6000,)
train: (2000, 10) (2000,)
train: (2000, 10) (2000,)


Применим OneHotEncoding для кодирования категориальных переменных.

Функция для кодирования категориальных переменных.

In [38]:
enc = OneHotEncoder(sparse=False, drop='first')
enc_2 = OneHotEncoder(sparse=False, drop='first')

enc.fit(features_train[['geography']])
enc_2.fit(features_train[['gender']])

def ohe_coding(features, encoder, encoder_2):
    
    enc = pd.DataFrame(encoder.transform(features[['geography']]), 
             columns=encoder.categories_[0][1:], index=features.index)
    features = features.join(enc).drop('geography', axis=1)


    enc_2 = pd.DataFrame(encoder_2.transform(features[['gender']]), 
             columns=encoder_2.categories_[0][1:], index=features.index)
    features = features.join(enc_2).drop('gender', axis=1)
    return features

In [39]:
features_train = ohe_coding(features_train, enc, enc_2)
features_test = ohe_coding(features_test, enc, enc_2)
features_valid = ohe_coding(features_valid, enc, enc_2)
features_train

credit_score  age  tenure    balance  num_of_products  has_cr_card  \
7479           567   35       8  153137.74                1            1   
3411           711   37       8  113899.92                1            0   
6027           850   44       3  140393.65                2            0   
1247           512   24       6       0.00                2            1   
3716           665   27       2  147435.96                1            0   
...            ...  ...     ...        ...              ...          ...   
4478           549   31       4       0.00                2            0   
4094           513   29      10       0.00                2            0   
3492           655   45       5   57327.04                1            0   
2177           667   24       4       0.00                2            0   
4578           691   38       1       0.00                2            0   

      is_active_member  estimated_salary  Germany  Spain  Male  
7479                 0          88659.07      0.0    1.0   1.0  
3411                 0          80215.20      0.0    0.0   0.0  
6027                 1         186285.52      1.0    0.0   1.0  
1247                 0          37654.31      0.0    0.0   1.0  
3716                 0         187508.06      1.0    0.0   0.0  
...                ...               ...      ...    ...   ...  
4478                 1          25684.85      0.0    0.0   1.0  
4094                 1          25514.77      0.0    0.0   1.0  
3492                 1          47349.00      0.0    0.0   0.0  
2177                 1          34335.55      0.0    0.0   0.0  
4578                 0          44653.50      0.0    0.0   1.0  

[6000 rows x 11 columns]

### Стандартизация данных

In [92]:
scaler = StandardScaler()
numeric = ['credit_score', 'age', 'tenure', 'balance', 'num_of_products', 'estimated_salary']
scaler.fit(features_train[numeric])
features_train[numeric] = scaler.transform(features_train[numeric])
features_train

credit_score       age    tenure   balance  num_of_products  \
7479     -0.886751 -0.373192  1.082277  1.232271        -0.891560   
3411      0.608663 -0.183385  1.082277  0.600563        -0.891560   
6027      2.052152  0.480939 -0.737696  1.027098         0.830152   
1247     -1.457915 -1.417129  0.354288 -1.233163         0.830152   
3716      0.130961 -1.132419 -1.101690  1.140475        -0.891560   
...            ...       ...       ...       ...              ...   
4478     -1.073677 -0.752805 -0.373701 -1.233163         0.830152   
4094     -1.447531 -0.942612  1.810266 -1.233163         0.830152   
3492      0.027113  0.575842 -0.009707 -0.310229        -0.891560   
2177      0.151731 -1.417129 -0.373701 -1.233163         0.830152   
4578      0.400966 -0.088482 -1.465685 -1.233163         0.830152   

      has_cr_card  is_active_member  estimated_salary  Germany  Spain  Male  
7479            1                 0         -0.187705      0.0    1.0   1.0  
3411            0                 0         -0.333945      0.0    0.0   0.0  
6027            0                 1          1.503095      1.0    0.0   1.0  
1247            1                 0         -1.071061      0.0    0.0   1.0  
3716            0                 0          1.524268      1.0    0.0   0.0  
...           ...               ...               ...      ...    ...   ...  
4478            0                 1         -1.278361      0.0    0.0   1.0  
4094            0                 1         -1.281307      0.0    0.0   1.0  
3492            0                 1         -0.903158      0.0    0.0   0.0  
2177            0                 1         -1.128539      0.0    0.0   0.0  
4578            0                 0         -0.949841      0.0    0.0   1.0  

[6000 rows x 11 columns]

In [93]:
features_test[numeric] = scaler.transform(features_test[numeric])
features_valid[numeric] = scaler.transform(features_valid[numeric])

Промасштабировали признаки.

### Обучение без учета дисбаланса классов

#### Дерево решений

In [94]:
best_model = None
best_result = 0
for depth in range(1, 16):
    model = DecisionTreeClassifier(max_depth=depth, criterion='entropy', splitter='best', 
                               min_samples_split=150, min_samples_leaf=10, random_state=12345)
    model.fit(features_train, target_train)
    prediction_valid = model.predict(features_valid)
    result = f1_score(target_valid, prediction_valid)
    print('depth:', depth, '| result:', result)
    
    if result > best_result:
        best_model = model
        best_result = result
        best_depth = depth
        auc_roc = roc_auc_score(target_valid, model.predict_proba(features_valid)[:, 1])

print('Лучшая модель: depth:', best_depth, '| result:', best_result, '| auc_roc:', auc_roc)

depth: 1 | result: 0.0
depth: 2 | result: 0.4986301369863014
depth: 3 | result: 0.3957219251336898
depth: 4 | result: 0.4320557491289198
depth: 5 | result: 0.5214814814814815
depth: 6 | result: 0.5185185185185186
depth: 7 | result: 0.5476190476190476
depth: 8 | result: 0.5476190476190476
depth: 9 | result: 0.5476190476190476
depth: 10 | result: 0.5476190476190476
depth: 11 | result: 0.5476190476190476
depth: 12 | result: 0.5476190476190476
depth: 13 | result: 0.5476190476190476
depth: 14 | result: 0.5476190476190476
depth: 15 | result: 0.5476190476190476
Лучшая модель: depth: 7 | result: 0.5476190476190476 | auc_roc: 0.8342177969061764


Максимум F1-меры, которой удалось добиться на валидационной выборке, изменяя гиперпараметры - 0.5476. Этого не достаточно.

AUC-ROC = 0.834 - хороший показатель, больше 0.59.

#### Случайный лес

In [95]:
best_model = None
best_result = 0
for estimators in range(5, 120, 5):
    for depth in range(7, 15):
        model = RandomForestClassifier(max_depth=depth, n_estimators=estimators, criterion='gini', 
                               min_samples_split=100, min_samples_leaf=10, random_state=12345)
        model.fit(features_train, target_train)
        prediction_valid = model.predict(features_valid)
        result = f1_score(target_valid, prediction_valid)
    
        if result > best_result:
            best_model = model
            best_result = result
            best_n = estimators
            best_depth = depth
            auc_roc = roc_auc_score(target_valid, model.predict_proba(features_valid)[:, 1])
print('Лучшая модель: depth:', depth, '| n_estimators:', best_n, '| result:', best_result, '| auc_roc:', auc_roc)


Лучшая модель: depth: 14 | n_estimators: 5 | result: 0.5117004680187208 | auc_roc: 0.8371005485173242


Максимум F1-меры, которой удалось добиться на валидационной выборке, изменяя гиперпараметры - 0.5117. Этого не достаточно.

AUC-ROC = 0.837 - хороший показатель, больше 0.59.

#### Логистическая регрессия

In [96]:
model_lr = LogisticRegression(penalty='l2' , solver='liblinear', random_state=12345)
model_lr.fit(features_train, target_train)
prediction_valid = model_lr.predict(features_valid)
result = f1_score(target_valid, prediction_valid)
auc_roc = roc_auc_score(target_valid, model_lr.predict_proba(features_valid)[:, 1])

print('F1-мера модели Логистической регрессии:', result, '| auc_roc:', auc_roc)

F1-мера модели Логистической регрессии: 0.27478260869565213 | auc_roc: 0.7386410142248726


Максимум F1-меры, которой удалось добиться на валидационной выборке, изменяя гиперпараметры - 0.2748. Этого не достаточно.

AUC-ROC = 0.738 - хороший показатель, больше 0.59, но значительно хуже предыдущих моделей.

### Учтем дисбаланс классов

In [97]:
data['exited'].value_counts()

0    7963
1    2037
Name: exited, dtype: int64

In [98]:
target_train.value_counts()

0    4804
1    1196
Name: exited, dtype: int64

#### Увеличение выборки

Увеличим количество положительных классов в 4 раза.

In [99]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]
    
    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345)
    
    return features_upsampled, target_upsampled

features_upsampled, target_upsampled = upsample(features_train, target_train, 4)

target_upsampled.value_counts()

0    4804
1    4784
Name: exited, dtype: int64

##### Дерево решений

In [100]:
best_model = None
best_result = 0
for depth in range(1, 16):
    model = DecisionTreeClassifier(max_depth=depth, criterion='entropy', splitter='best', 
                               min_samples_split=150, min_samples_leaf=10, random_state=12345)
    model.fit(features_upsampled, target_upsampled)
    prediction_valid = model.predict(features_valid)
    result = f1_score(target_valid, prediction_valid)
    print('depth:', depth, '| result:', result)
    
    if result > best_result:
        best_model = model
        best_result = result
        best_depth = depth
        auc_roc = roc_auc_score(target_valid, model.predict_proba(features_valid)[:, 1])

print('Лучшая модель: depth:', best_depth, '| result:', best_result, '| auc_roc:', auc_roc)

depth: 1 | result: 0.48514851485148514
depth: 2 | result: 0.5129682997118156
depth: 3 | result: 0.5129682997118156
depth: 4 | result: 0.5388349514563108
depth: 5 | result: 0.5848074921956294
depth: 6 | result: 0.5711775043936732
depth: 7 | result: 0.5803402646502835
depth: 8 | result: 0.5826170009551098
depth: 9 | result: 0.5833333333333334
depth: 10 | result: 0.5917753259779338
depth: 11 | result: 0.5876591576885407
depth: 12 | result: 0.5870841487279844
depth: 13 | result: 0.5870841487279844
depth: 14 | result: 0.5870841487279844
depth: 15 | result: 0.5870841487279844
Лучшая модель: depth: 10 | result: 0.5917753259779338 | auc_roc: 0.815772234139994


Максимум F1-меры, которой удалось добиться на валидационной выборке, изменяя гиперпараметры - 0.5917. Нам этого достаточно, продолжим исследование на других моделях.

AUC-ROC = 0.8158 - хороший показатель, больше 0.59.

##### Случайный лес

In [101]:
best_model = None
best_result = 0
for estimators in range(5, 120, 5):
    for depth in range(7, 15):
        model = RandomForestClassifier(max_depth=depth, n_estimators=estimators, criterion='gini', 
                               min_samples_split=150, min_samples_leaf=10, random_state=12345)
        model.fit(features_upsampled, target_upsampled)
        prediction_valid = model.predict(features_valid)
        result = f1_score(target_valid, prediction_valid)
    
        if result > best_result:
            best_model = model
            best_result = result
            best_n = estimators
            best_depth = depth
            auc_roc = roc_auc_score(target_valid, model.predict_proba(features_valid)[:, 1])
print('Лучшая модель: depth:', depth, '| n_estimators:', best_n, '| result:', best_result, '| auc_roc:', auc_roc)

Лучшая модель: depth: 14 | n_estimators: 20 | result: 0.614 | auc_roc: 0.8561997148729296


Максимум F1-меры, которой удалось добиться на валидационной выборке, изменяя гиперпараметры - 0.614. Нам этого достаточно, продолжим исследование на других моделях.

AUC-ROC = 0.8562 - хороший показатель, больше 0.59.

##### Логистическая регрессия

In [102]:
model_lr = LogisticRegression(penalty='l2' , solver='liblinear', random_state=12345)
model_lr.fit(features_upsampled, target_upsampled)
prediction_valid = model_lr.predict(features_valid)
result = f1_score(target_valid, prediction_valid)
auc_roc = roc_auc_score(target_valid, model_lr.predict_proba(features_valid)[:, 1])

print('F1-мера модели Логистической регрессии:', result, '| auc_roc:', auc_roc)

F1-мера модели Логистической регрессии: 0.47619047619047616 | auc_roc: 0.7417621212734475


Максимум F1-меры, которой удалось добиться на валидационной выборке, изменяя гиперпараметры - 0.4762. Этого не достаточно.

AUC-ROC = 0.7418 - хороший показатель, больше 0.59.

#### Уменьшение выборки

In [103]:
def downsample(features, target, fraction):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_downsampled = pd.concat(
        [features_zeros.sample(frac=fraction, random_state=12345)] + [features_ones])
    target_downsampled = pd.concat(
        [target_zeros.sample(frac=fraction, random_state=12345)] + [target_ones])
    
    features_downsampled, target_downsampled = shuffle(
        features_downsampled, target_downsampled, random_state=12345)
    
    return features_downsampled, target_downsampled

features_downsampled, target_downsampled = downsample(features_train, target_train, 0.25)

target_downsampled.value_counts()

0    1201
1    1196
Name: exited, dtype: int64

Получили маленькие выборки, проверим на моделях.

In [104]:
best_model = None
best_result = 0
for depth in range(1, 16):
    model = DecisionTreeClassifier(max_depth=depth, criterion='entropy', splitter='best', 
                               min_samples_split=150, min_samples_leaf=10, random_state=12345)
    model.fit(features_downsampled, target_downsampled)
    prediction_valid = model.predict(features_valid)
    result = f1_score(target_valid, prediction_valid)
    print('depth:', depth, '| result:', result)
    
    if result > best_result:
        best_model = model
        best_result = result
        best_depth = depth
        auc_roc = roc_auc_score(target_valid, model.predict_proba(features_valid)[:, 1])

print('Лучшая модель: depth:', best_depth, '| result:', best_result, '| auc_roc:', auc_roc)

depth: 1 | result: 0.4878957169459963
depth: 2 | result: 0.5109090909090909
depth: 3 | result: 0.5611510791366906
depth: 4 | result: 0.5557851239669421
depth: 5 | result: 0.5590622182146078
depth: 6 | result: 0.5550883095037847
depth: 7 | result: 0.5476815398075241
depth: 8 | result: 0.5476815398075241
depth: 9 | result: 0.5476815398075241
depth: 10 | result: 0.5476815398075241
depth: 11 | result: 0.5476815398075241
depth: 12 | result: 0.5476815398075241
depth: 13 | result: 0.5476815398075241
depth: 14 | result: 0.5476815398075241
depth: 15 | result: 0.5476815398075241
Лучшая модель: depth: 3 | result: 0.5611510791366906 | auc_roc: 0.7929643171416536


Получили слишком низкое значение F-1 меры - 0.561. AUC-ROC = 0.793. 

In [105]:
best_model = None
best_result = 0
for estimators in range(5, 120, 5):
    for depth in range(7, 15):
        model = RandomForestClassifier(max_depth=depth, n_estimators=estimators, criterion='gini', 
                               min_samples_split=150, min_samples_leaf=10, random_state=12345)
        model.fit(features_downsampled, target_downsampled)
        prediction_valid = model.predict(features_valid)
        result = f1_score(target_valid, prediction_valid)
    
        if result > best_result:
            best_model = model
            best_result = result
            best_n = estimators
            best_depth = depth
            auc_roc = roc_auc_score(target_valid, model.predict_proba(features_valid)[:, 1])
print('Лучшая модель: depth:', depth, '| n_estimators:', best_n, '| result:', best_result, '| auc_roc:', auc_roc)

Лучшая модель: depth: 14 | n_estimators: 10 | result: 0.5874524714828897 | auc_roc: 0.8438494553053573


F1-мера = 0.5875, это ниже требуемого значения. AUC-ROC = 0.77

In [106]:
model_lr = LogisticRegression(penalty='l2' , solver='liblinear', random_state=12345)
model_lr.fit(features_downsampled, target_downsampled)
prediction_valid = model_lr.predict(features_valid)
result = f1_score(target_valid, prediction_valid)
auc_roc = roc_auc_score(target_valid, model_lr.predict_proba(features_valid)[:, 1])

print('F1-мера модели Логистической регрессии:', result, '| auc_roc:', auc_roc)

F1-мера модели Логистической регрессии: 0.48417450812660395 | auc_roc: 0.7413123940330189


Получили слишком низкие показатели метрик

#### Изменение порога

In [107]:
model_lr = LogisticRegression(penalty='l2' , solver='liblinear', class_weight='balanced' , random_state=12345)
model_lr.fit(features_train, target_train)
prediction_valid = model_lr.predict(features_valid)
result = f1_score(target_valid, prediction_valid)
auc_roc = roc_auc_score(target_valid, model_lr.predict_proba(features_valid)[:, 1])

print('F1-мера модели Логистической регрессии:', result, '| auc_roc:', auc_roc)

F1-мера модели Логистической регрессии: 0.4788245462402766 | auc_roc: 0.7418085930882918


Получили слишком низкие показатель F-1 меры = 0.4788. AUC-ROC = 0.742

## Тестирование лучшей модели

Лучшая подобранная модель - случайный лес, обученный на увеличенной выборке.
- RandomForestClassifier(max_depth=14, n_estimators=20, criterion='gini', min_samples_split=150, min_samples_leaf=10, random_state=12345)

In [108]:
model = RandomForestClassifier(max_depth=14, n_estimators=20, criterion='gini', 
                               min_samples_split=150, min_samples_leaf=10, random_state=12345)
model.fit(features_upsampled, target_upsampled)
prediction_test = model.predict(features_test)
result = f1_score(target_test, prediction_test)
auc_roc = roc_auc_score(target_valid, model.predict_proba(features_valid)[:, 1])
print('F1 =', result, '| AUC-ROC =', auc_roc)

F1 = 0.6136595310907238 | AUC-ROC = 0.8503615057467647


Вывод: в данной работе была выполнена первоначальная подготовка данных к машинному обучению: оптимизированы типы данных, заполнены пропуски, удалены лишние столбцы. Затем происходил подбор наилучшей модели, опираясь на F1-меру и параллельно рассчитывая AUC-ROC (сначала без балансировки классов, затем увеличив выборку). По итогу лучшая подобранная модель - случайный лес, обученный на увеличенной выборке. Удалось достичь на тестовой выборке следующих показателей: F1-мера = 0.613659, AUC-ROC = 0.85

Итоговый вывод:

Была проделана объемная работа, цель которой - спрогнозировать, уйдет ли клиент из банка в ближайшее время. Началось всё с подготовки данных к машинному обучению:

* пропуски были заполнены медианой
* удалены лишние столбцы
* изменены типы данных в датафрейме, для оптимального использования памяти

Далее я разделил датафрейм на признаки и целевой признак, разделил всё на 3 выборки (обучающую, валидационную, тестовую) и  принялся за кодирование категориальных признаков методом OHE. После провел масштабирование численных признаков.

Следующий этап - построение прогнозирующих моделей. Использовались модели следующих типов: дерево решений, случайный лес, логистическая регрессия. Лучшие результаты показывала модель решающего дерева. Далее я стал учитывать дисбаланс классов двумя способами: увеличение выборки и уменьшение выборки. Как показала практика - уменьшение выборки в этой работе не дало хороших результатов. Наоборот, увеличение выборки позволило существенно повысить качество моделей. При таком раскладе лучшей была модель случайного леса со следующими гиперпараметрами: `max_depth=14`, `n_estimators=20`, `criterion='gini'`, `min_samples_split=150`, `min_samples_leaf=10`, `random_state=12345`.
В итоге на тестовой выборке удалось получить следующие результаты: F1-мера = 0.613659, AUC-ROC = 0.85